File: Desafio Mobile iOS/Desafio Mobile iOS/Controller/PullRequestTableViewController.m
//
//  PullRequestTableViewController.m
//  Desafio Mobile iOS
//
//  Created by Adriano Rezena on 12/10/17.
//  Copyright © 2017 Adriano Rezena. All rights reserved.
//

#import "PullRequestTableViewController.h"
#import "PullRequestTableViewCell.h"
#import "LibraryAPI.h"

@interface PullRequestTableViewController () {
    NSArray *pullRequestsArray;
    UIActivityIndicatorView *activityIndicatorView;
}

@end

static Repository *repository;

@implementation PullRequestTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.title = repository.name;
    [self AtualizarView];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(AtualizarView) name:kStrNotificationPullRequestFinished object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(notificarPullRequestError:) name:kStrNotificationPullRequestError object:nil];
    
    activityIndicatorView = [UIActivityIndicatorView new];
    activityIndicatorView.frame = CGRectMake(0, 0, 20, 20);
    activityIndicatorView.center = self.tableView.center;
    activityIndicatorView.activityIndicatorViewStyle = UIActivityIndicatorViewStyleGray;
    [activityIndicatorView startAnimating];
    [self.tableView addSubview:activityIndicatorView];
    
    [[LibraryAPI sharedInstance] getPullRequestsfromRepository:repository];
}


- (void)setRepository:(Repository *) repo {
    repository = repo;
}


-(void)notificarPullRequestError:(NSNotification *)notification {
    NSString *mensagem = notification.userInfo[@"mensagem"];
    
    UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Falha ao atualizar" message:mensagem delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
    [alert show];
    
    [activityIndicatorView stopAnimating];
    [activityIndicatorView setHidden:YES];
}


-(void)AtualizarView {
    [activityIndicatorView stopAnimating];
    [activityIndicatorView setHidden:YES];
    
    NSMutableArray *pullRequestsMutableArray = [NSMutableArray new];
    pullRequestsArray = [NSArray new];
    
    RLMRealm *realm = [RLMRealm defaultRealm];
    RLMResults<PullRequest *> *pullRequests = [PullRequest objectsWhere:@"repositoryId = %i", repository.id];
    
    [realm transactionWithBlock:^{
        for (PullRequest *pull in pullRequests) {
            [pullRequestsMutableArray addObject:pull];
        }
    }];
    
    pullRequestsArray = [pullRequestsMutableArray copy];
    [self.tableView reloadData];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [pullRequestsArray count];
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    PullRequestTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"pullRequestCell" forIndexPath:indexPath];
    
    PullRequest *pullRequest = pullRequestsArray[indexPath.row];
    [cell configureWith:pullRequest];

    return cell;
}

-(CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 130;
}


-(UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    
    if ([pullRequestsArray count] > 0) {
        UIView *view = [[UIView alloc] initWithFrame:CGRectMake(0, 0, tableView.frame.size.width, 30)];
        [view setBackgroundColor:[UIColor whiteColor]];
        
        int openedCount = 0;
        int closedCount = 0;
        
        for (PullRequest *pullRequest in pullRequestsArray) {
            if ([pullRequest.state isEqualToString:@"open"]) {
                openedCount++;
            } else if ([pullRequest.state isEqualToString:@"closed"]) {
                closedCount++;
            }
        }
        
        UILabel *openedLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, view.frame.size.width / 2, view.frame.size.height)];
        openedLabel.textColor = [UIColor orangeColor];
        openedLabel.font = [UIFont boldSystemFontOfSize:13];
        openedLabel.textAlignment = NSTextAlignmentRight;
        openedLabel.text = [NSString stringWithFormat:@"%i opened", openedCount];
        [view addSubview:openedLabel];
        
        UILabel *closedLabel = [[UILabel alloc] initWithFrame:CGRectMake(view.frame.size.width / 2, 0, view.frame.size.width / 2, view.frame.size.height)];
        closedLabel.textColor = [UIColor blackColor];
        closedLabel.font = [UIFont boldSystemFontOfSize:13];
        closedLabel.text = [NSString stringWithFormat:@"/ %i closed", closedCount];
        [view addSubview:closedLabel];
        
        return view;
    } else {
        return nil;
    }
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    PullRequest *pullRequest = pullRequestsArray[indexPath.row];
    
    NSURL *url = [NSURL URLWithString:pullRequest.url];
    [[UIApplication sharedApplication] openURL:url];
}

@end

File: GitHubAllStars/GitHubAllStars/ViewControllers/PullRequestTableViewController.swift
//
//  PullRequestTableViewController.swift
//  GitHubAllStars
//
//  Created by Raphael Araújo on 23/10/17.
//  Copyright © 2017 Raphael Araújo. All rights reserved.
//

import UIKit
class PullRequestTableViewController: UITableViewController {
    var repository: Repository?
    var pullRequests = [PullRequest]()
    override func viewDidLoad() {
        super.viewDidLoad()

        self.navigationItem.title = self.repository?.name
        let nib = UINib(nibName: "PullRequestTableViewCell", bundle: nil)
        self.tableView.register(nib, forCellReuseIdentifier:
            "PullRequestTableViewCell")
        self.tableView.rowHeight = UITableViewAutomaticDimension
        self.tableView.estimatedRowHeight = 143
        self.refreshControl?.addTarget(self, action: #selector(handleRefresh(_:)), for: UIControlEvents.valueChanged)
        self.refreshControl?.tintColor = .gray
        self.tableView.addSubview(self.refreshControl!)
        
        self.loadData()
    }

    @objc func handleRefresh(_: UIRefreshControl) {
        
        loadData()
    }
    
    // MARK: - Table view data source
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.pullRequests.count
    }
    
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "PullRequestTableViewCell", for: indexPath) as! PullRequestTableViewCell
        cell.pullRequest = self.pullRequests[indexPath.row]
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let pullRequest = self.pullRequests[indexPath.row]
        UIApplication.shared.openURL(URL(string: pullRequest.html_url)!)
    }
    
    func loadData() {
        if let fullRepositoryName = self.repository?.fullName {
            DataManager.sharedInstance.getPullRequest(repository: fullRepositoryName) { (error, message, pullRequestsResponse) in
                if(!error) {
                    if let prs = pullRequestsResponse {
                        self.pullRequests.removeAll()
                        self.pullRequests.append(contentsOf: prs)
                        print("PRS: \(self.pullRequests.count)")
                        self.tableView.reloadData()
                    }
                }
                self.refreshControl?.endRefreshing()
            }
        }
    }
}
